komodo/check_up_to_date_pypi.py

```diff
@@ -0,0 +1,167 @@
+import copy
+import sys
+import pathlib
+
+from packaging import version as get_version
+from packaging.specifiers import SpecifierSet
+import requests
+import ruamel.yaml
+import argparse
+from komodo.package_version import LATEST_PACKAGE_ALIAS, strip_version
+from komodo.prettier import write_to_file
+
+
+def yaml_parser():
+    yaml = ruamel.yaml.YAML()
+    yaml.indent(mapping=2, sequence=4, offset=2)
+    yaml.preserve_quotes = True
+    return yaml
+
+
+def get_release(yaml, fname):
+    with open(fname, "r") as fin:
+        result = yaml.load(fin)
+    return result
+
+
+def get_packages_pypi(package_names):
+    return [
+        (package, requests.get(f"https://pypi.python.org/pypi/{package}/json"))
+        for package in package_names
+    ]
+
+
+def get_python_requirement(sources: list):
+    """
+    Goes through the different sources of the package and returns the first
+    python requirement
+    :param sources:
+    :return: Python requirement or empty string if there is no requirement
+    """
+    for source in sources:
+        if source.get("yanked", False):
+            continue
+        required_python = source.get("requires_python", "")
+        if required_python:
+            return required_python
+    return ""
+
+
+def get_compatible_versions(releases: dict, python_version):
+    compatible_versions = []
+    for version_str, build_info in releases.items():
+        package_version = get_version.parse(version_str)
+        if package_version.is_prerelease:
+            continue
+        required_python = SpecifierSet(get_python_requirement(build_info))
+        if python_version in required_python:
+            compatible_versions.append(package_version)
+    return compatible_versions
+
+
+def evaluate_response(response: dict, python_version: str):
+    python_requirement = response["info"]["requires_python"]
+    if not python_requirement or python_version in SpecifierSet(python_requirement):
+        return get_version.parse(response["info"]["version"])
+    return max(get_compatible_versions(response["releases"], python_version))
+
+
+def get_pypi_packages(release: dict, repository: dict) -> list:
+    pypi_packages = []
+    for package, version in release.items():
+        if repository[package][version].get("source", None) == "pypi":
+            pypi_packages.append(package)
+    return pypi_packages
+
+
+def get_upgrade_proposals(release_file, repository_file, python_version, propose_upgrade):
+    yaml = yaml_parser()
+    releases = get_release(yaml, release_file)
+    repository = get_release(yaml, repository_file)
+
+    pypi_packages = get_pypi_packages(releases, repository)
+    pypi_response = get_packages_pypi(pypi_packages)
```

Comment:
The naming here is slightly confusing, we could call the bottom one get_pypi_info or something like that instead?